// @opra/core/cjs/http/http-server-response.js
"use strict";
/*
  Some parts of this file contains codes from open source express library
  https://github.com/expressjs
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpServerResponse = void 0;
const tslib_1 = require("tslib");
const content_disposition_1 = tslib_1.__importDefault(require("content-disposition"));
const content_type_1 = tslib_1.__importDefault(require("content-type"));
const cookie_1 = tslib_1.__importDefault(require("cookie"));
const cookie_signature_1 = tslib_1.__importDefault(require("cookie-signature"));
const encodeurl_1 = tslib_1.__importDefault(require("encodeurl"));
const mime_types_1 = tslib_1.__importDefault(require("mime-types"));
const path_1 = tslib_1.__importDefault(require("path"));
const putil_varhelpers_1 = require("putil-varhelpers");
const vary_1 = tslib_1.__importDefault(require("vary"));
const common_1 = require("@opra/common");
const http_outgoing_message_host_js_1 = require("./impl/http-outgoing-message.host.js");
const charsetRegExp = /;\s*charset\s*=/;
function isHttpIncomingMessage(v) {
    return v && typeof v.getHeaders === 'function' && (0, common_1.isStream)(v);
}
var HttpServerResponse;
(function (HttpServerResponse) {
    function from(instance) {
        if (!isHttpIncomingMessage(instance))
            instance = new http_outgoing_message_host_js_1.HttpOutgoingMessageHost(instance);
        (0, common_1.mergePrototype)(instance, HttpServerResponseHost.prototype);
        return instance;
    }
    HttpServerResponse.from = from;
})(HttpServerResponse || (exports.HttpServerResponse = HttpServerResponse = {}));
class HttpServerResponseHost {
    attachment(filename) {
        if (filename)
            this.contentType(path_1.default.extname(filename));
        this.setHeader('Content-Disposition', (0, content_disposition_1.default)(filename));
        return this;
    }
    contentType(type) {
        const ct = type.indexOf('/') === -1
            ? mime_types_1.default.lookup(type)
            : type;
        this.setHeader('Content-Type', ct);
        return this;
    }
    setHeader(field, val) {
        const setHeader = Object.getPrototypeOf(this).setHeader;
        if (typeof field === 'object') {
            for (const [k, v] of Object.entries(field)) {
                this.setHeader(k, v);
            }
            return this;
        }
        const fieldLower = field.toLowerCase();
        let value = Array.isArray(val)
            ? val.map(String)
            : (val ? String(val) : '');
        // add charset to content-type
        if (fieldLower === 'content-type') {
            if (Array.isArray(value)) {
                throw new TypeError('Content-Type cannot be set to an Array');
            }
            if (!charsetRegExp.test(value)) {
                const charset = mime_types_1.default.charsets.lookup(value.split(';')[0]);
                if (charset)
                    value += '; charset=' + charset.toLowerCase();
            }
        }
        setHeader.call(this, field, value);
        return this;
    }
    clearCookie(name, options) {
        const opts = {
            expires: new Date(1),
            path: '/',
            ...options
        };
        return this.cookie(name, '', opts);
    }
    cookie(name, value, options) {
        const opts = { ...options };
        let val = typeof value === 'object'
            ? 'j:' + JSON.stringify(value)
            : String(value);
        if (opts.signed) {
            const secret = opts.secret || this.req?.secret;
            if (!secret)
                throw new Error('"secret" required for signed cookies');
            val = 's:' + cookie_signature_1.default.sign(val, secret);
        }
        if (opts.maxAge != null) {
            const maxAge = opts.maxAge - 0;
            if (!isNaN(maxAge)) {
                opts.expires = new Date(Date.now() + maxAge);
                opts.maxAge = Math.floor(maxAge / 1000);
            }
        }
        if (opts.path == null)
            opts.path = '/';
        this.appendHeader('Set-Cookie', cookie_1.default.serialize(name, String(val), opts));
        return this;
    }
    status(code) {
        this.statusCode = code;
        return this;
    }
    sendStatus(statusCode) {
        const body = common_1.HttpStatusCode[statusCode] || String(statusCode);
        this.statusCode = statusCode;
        this.contentType('txt');
        return this.send(body);
    }
    links(links) {
        let link = this.getHeader('Link') || '';
        if (link)
            link += ', ';
        this.setHeader('Link', link +
            Object.keys(links)
                .map(rel => '<' + links[rel] + '>; rel="' + rel + '"')
                .join(', '));
        return this;
    }
    location(url) {
        let loc = url;
        // "back" is an alias for the referrer
        if (url === 'back')
            loc = this.req?.get('Referrer') || '/';
        // set location
        return this.setHeader('Location', (0, encodeurl_1.default)(loc));
    }
    redirect(arg0, arg1) {
        const address = String(arg1 || arg0);
        const status = typeof arg0 === 'number' ? arg0 : 302;
        // Set location header
        this.location(address);
        // Respond
        this.statusCode = status;
        this.end();
    }
    send(body) {
        let chunk = body;
        let encoding;
        const req = this.req;
        let ctype = (0, putil_varhelpers_1.toString)(this.getHeader('Content-Type'));
        if (typeof chunk !== 'string') {
            if (chunk === null)
                chunk = '';
            else if (Buffer.isBuffer(chunk)) {
                if (!ctype)
                    this.contentType('bin');
            }
            else {
                ctype = 'json';
                chunk = JSON.stringify(chunk);
            }
        }
        // write strings in utf-8
        if (typeof chunk === 'string') {
            encoding = 'utf-8';
            this.setHeader('Content-Type', setCharset(ctype || 'txt', encoding));
        }
        // populate Content-Length
        let len = 0;
        if (chunk !== undefined) {
            if (Buffer.isBuffer(chunk)) {
                // get length of Buffer
                len = chunk.length;
            }
            else if (chunk.length < 1000) {
                // just calculate length when small chunk
                len = Buffer.byteLength(chunk, encoding);
            }
            else {
                // convert chunk to Buffer and calculate
                chunk = Buffer.from(chunk, encoding);
                encoding = undefined;
                len = chunk.length;
            }
            this.setHeader('Content-Length', len);
        }
        // freshness
        if (req?.fresh)
            this.statusCode = 304;
        // strip irrelevant headers
        if (204 === this.statusCode || 304 === this.statusCode) {
            this.removeHeader('Content-Type');
            this.removeHeader('Content-Length');
            this.removeHeader('Transfer-Encoding');
            chunk = '';
        }
        // alter headers for 205
        if (this.statusCode === 205) {
            this.setHeader('Content-Length', '0');
            this.removeHeader('Transfer-Encoding');
            chunk = '';
        }
        if (req?.method === 'HEAD') {
            // skip body for HEAD
            this.end();
        }
        else {
            // respond
            if (encoding)
                this.end(chunk, encoding);
            else
                this.end(chunk);
        }
        return this;
    }
    vary(field) {
        (0, vary_1.default)(this, field);
        return this;
    }
}
function setCharset(type, charset) {
    if (!(type && charset))
        return type;
    // parse type
    const parsed = content_type_1.default.parse(type);
    // set charset
    parsed.parameters.charset = charset;
    // format type
    return content_type_1.default.format(parsed);
}
