// @opra/common/cjs/document/resource/action.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Action = void 0;
const valgen_1 = require("valgen");
const data_type_js_1 = require("../data-type/data-type.js");
const endpoint_js_1 = require("./endpoint.js");
/**
 *
 * @class Action
 */
class Action extends endpoint_js_1.Endpoint {
    constructor(resource, name, init) {
        super(resource, name, init);
        this.resource = resource;
        this.name = name;
        this.kind = 'action';
        this.encodeReturning = valgen_1.isAny;
        if (init.returnType)
            this.returnType = init.returnType instanceof data_type_js_1.DataType
                ? init.returnType : this.resource.document.getDataType(init.returnType);
        this.returnMime = init.returnMime;
    }
    exportSchema(options) {
        const schema = super.exportSchema(options);
        if (this.returnType)
            schema.returnType = this.returnType.isEmbedded
                ? this.returnType.exportSchema(options)
                : this.returnType.name;
        if (this.returnMime)
            schema.returnMime = this.returnMime;
        return schema;
    }
}
exports.Action = Action;
