// @opra/common/cjs/document/resource/container-class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerClass = void 0;
const index_js_1 = require("../../exception/index.js");
const index_js_2 = require("../../helpers/index.js");
const index_js_3 = require("../../schema/index.js");
const resource_js_1 = require("./resource.js");
const PATH_PREFIX_PATTERN = /^(\/*)(.+)$/;
class ContainerClass extends resource_js_1.Resource {
    constructor(owner, init) {
        super(owner instanceof ContainerClass ? owner.document : owner, init);
        this.kind = index_js_3.OpraSchema.Container.Kind;
        this.resources = new index_js_2.ResponsiveMap();
        this.parent = owner instanceof ContainerClass ? owner : undefined;
    }
    exportSchema(options) {
        const schema = super.exportSchema(options);
        if (this.resources.size) {
            const resources = schema.resources = {};
            for (const [name, r] of this.resources.entries()) {
                resources[name] = r.exportSchema(options);
            }
        }
        return schema;
    }
    getResource(path, silent) {
        let resource;
        path = PATH_PREFIX_PATTERN.exec(path)?.[2] || path;
        if (path.includes('/')) {
            const arr = path.split('/');
            let i;
            const l = arr.length;
            let container = this;
            for (i = 0; i < l; i++) {
                resource = container.resources.get(arr[i]);
                if (resource instanceof ContainerClass)
                    container = resource;
                else
                    break;
            }
            // If no resource found or walking through path not completed
            if (!resource || i < l - 1)
                resource = undefined;
        }
        else
            resource = this.resources.get(path);
        if (resource || silent)
            return resource;
        throw new index_js_1.ResourceNotAvailableError(path);
    }
    getContainer(path, silent) {
        const t = this.getResource(path);
        if (!t && silent)
            return;
        if (t && t.kind === index_js_3.OpraSchema.Container.Kind)
            return t;
        throw new index_js_1.NotAcceptableError(`Resource type "${t.name}" is not a Container`);
    }
    getCollection(path, silent) {
        const t = this.getResource(path);
        if (!t && silent)
            return;
        if (t && t.kind === index_js_3.OpraSchema.Collection.Kind)
            return t;
        throw new index_js_1.NotAcceptableError(`Resource type "${t.name}" is not a Collection`);
    }
    getSingleton(path, silent) {
        const t = this.getResource(path);
        if (!t && silent)
            return;
        if (t && t.kind === index_js_3.OpraSchema.Singleton.Kind)
            return t;
        throw new index_js_1.NotAcceptableError(`Resource type "${t.name}" is not a Singleton`);
    }
    getStorage(path, silent) {
        const t = this.getResource(path);
        if (!t && silent)
            return;
        if (t && t.kind === index_js_3.OpraSchema.Storage.Kind)
            return t;
        throw new index_js_1.NotAcceptableError(`Resource type "${t.name}" is not a Storage`);
    }
}
exports.ContainerClass = ContainerClass;
