// @opra/common/cjs/document/resource/parameter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Parameter = void 0;
const valgen_1 = require("valgen");
const index_js_1 = require("../../helpers/index.js");
/**
 *
 * @class Parameter
 */
class Parameter {
    constructor(name, init) {
        this.name = name;
        this.type = init.type;
        this.description = init.description;
        this.isArray = init.isArray;
        this.default = init.default;
        this.required = init.required;
        this.deprecated = init.deprecated;
        this.examples = init.examples;
        this.isBuiltin = init.isBuiltin;
    }
    exportSchema(options) {
        return (0, index_js_1.omitUndefined)({
            type: this.type.name ? this.type.name : this.type.exportSchema(options),
            description: this.description,
            isArray: this.isArray,
            default: this.default,
            required: this.required,
            deprecated: this.deprecated,
            examples: this.examples
        });
    }
    getDecoder() {
        if (!this._decoder)
            this._decoder = this.generateCodec('decode');
        return this._decoder;
    }
    getEncoder() {
        if (!this._encoder)
            this._encoder = this.generateCodec('encode');
        return this._encoder;
    }
    generateCodec(codec, options) {
        let fn = this.type.generateCodec(codec, options);
        if (this.isArray)
            fn = valgen_1.vg.stringSplit(',');
        return !options?.partial && this.required ? valgen_1.vg.required(fn) : valgen_1.vg.optional(fn);
    }
}
exports.Parameter = Parameter;
