// @opra/common/cjs/document/resource/resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Resource = void 0;
const index_js_1 = require("../../helpers/index.js");
const object_utils_js_1 = require("../../helpers/object-utils.js");
const inspect_util_js_1 = require("../utils/inspect.util.js");
const action_js_1 = require("./action.js");
class Resource {
    constructor(parent, init) {
        this.actions = new index_js_1.ResponsiveMap();
        if (parent instanceof Resource) {
            this.document = parent.document;
            this.parent = parent;
        }
        else
            this.document = parent;
        this.name = init.name;
        this.description = init.description;
        this.controller = init.controller;
        if (this.controller) {
            this.ctor = Object.getPrototypeOf(this.controller).constructor;
        }
        else
            this.ctor = init.ctor;
        if (init.actions) {
            for (const [name, meta] of Object.entries(init.actions)) {
                this.actions.set(name, new action_js_1.Action(this, name, meta));
            }
        }
    }
    getFullPath(documentPath) {
        if (this === this.document.root)
            return documentPath ? '/root' : '/';
        let out = this.parent?.getFullPath(documentPath);
        if (!out?.endsWith('/'))
            out += '/';
        return out + (documentPath ? 'resources/' : '') + this.name;
    }
    exportSchema(options) {
        const schema = (0, object_utils_js_1.omitUndefined)({
            kind: this.kind,
            description: this.description,
        });
        if (this.actions.size) {
            schema.actions = {};
            for (const action of this.actions.values()) {
                schema.actions[action.name] = action.exportSchema(options);
            }
        }
        return schema;
    }
    toString() {
        return `[${Object.getPrototypeOf(this).constructor.name} ${this.name || '#Embedded'}]`;
    }
    [inspect_util_js_1.nodeInspectCustom]() {
        return `[${inspect_util_js_1.colorFgYellow + Object.getPrototypeOf(this).constructor.name + inspect_util_js_1.colorReset}` +
            ` ${inspect_util_js_1.colorFgMagenta + this.name + inspect_util_js_1.colorReset}]`;
    }
}
exports.Resource = Resource;
