// @opra/common/cjs/document/resource/singleton-decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SingletonDecorator = void 0;
const index_js_1 = require("../../schema/index.js");
const action_decorator_js_1 = require("./action-decorator.js");
const collection_decorator_js_1 = require("./collection-decorator.js");
const resource_decorator_js_1 = require("./resource-decorator.js");
const operationProperties = ['create', 'delete', 'get', 'update'];
function SingletonDecorator(type, options) {
    return (0, resource_decorator_js_1.ResourceDecorator)(index_js_1.OpraSchema.Singleton.Kind, { ...options, type });
}
exports.SingletonDecorator = SingletonDecorator;
Object.assign(SingletonDecorator, resource_decorator_js_1.ResourceDecorator);
/*
 * Action PropertyDecorator
 */
(function (SingletonDecorator) {
    /**
     * Action PropertyDecorator
     */
    function Action(options) {
        const list = [];
        return (0, action_decorator_js_1.createActionDecorator)(options, operationProperties, list);
    }
    SingletonDecorator.Action = Action;
    function Create(options) {
        return collection_decorator_js_1.CollectionDecorator.Create(options);
    }
    SingletonDecorator.Create = Create;
})(SingletonDecorator || (exports.SingletonDecorator = SingletonDecorator = {}));
/*
 * Get PropertyDecorator
 */
(function (SingletonDecorator) {
    function Get(options) {
        return collection_decorator_js_1.CollectionDecorator.Get(options);
    }
    SingletonDecorator.Get = Get;
})(SingletonDecorator || (exports.SingletonDecorator = SingletonDecorator = {}));
/*
 * Delete PropertyDecorator
 */
(function (SingletonDecorator) {
    function Delete(options) {
        return collection_decorator_js_1.CollectionDecorator.Delete(options);
    }
    SingletonDecorator.Delete = Delete;
})(SingletonDecorator || (exports.SingletonDecorator = SingletonDecorator = {}));
/*
 * Update PropertyDecorator
 */
(function (SingletonDecorator) {
    function Update(options) {
        return collection_decorator_js_1.CollectionDecorator.Update(options);
    }
    SingletonDecorator.Update = Update;
})(SingletonDecorator || (exports.SingletonDecorator = SingletonDecorator = {}));
