// @opra/core/esm/http/adapters/node-http-adapter.host.js
import http from 'http';
import { HttpStatusCode, HttpStatusMessages, OpraURL, OpraURLPath, } from '@opra/common';
import { HttpAdapterHost } from '../http-adapter-host.js';
import { HttpServerRequest } from '../http-server-request.js';
import { HttpServerResponse } from '../http-server-response.js';
/**
 * @class NodeHttpAdapterHost
 */
export class NodeHttpAdapterHost extends HttpAdapterHost {
    constructor() {
        super(...arguments);
        this._platform = 'http';
    }
    get basePath() {
        return this._basePath;
    }
    get server() {
        return this._server;
    }
    async close() {
        await super.close();
        if (this.server.listening)
            await new Promise((resolve, reject) => {
                this.server.close((err) => {
                    if (err)
                        return reject(err);
                    resolve();
                });
            });
    }
    async init(api, options) {
        await super.init(api, options);
        this._basePath = new OpraURLPath(options?.basePath);
        this._server = http.createServer((incomingMessage, serverResponse) => this._serverListener(incomingMessage, serverResponse));
    }
    _serverListener(incomingMessage, serverResponse) {
        const originalUrl = incomingMessage.url;
        const parsedUrl = new OpraURL(originalUrl);
        const relativePath = OpraURLPath.relative(parsedUrl.path, this.basePath);
        if (!relativePath) {
            serverResponse.statusCode = HttpStatusCode.NOT_FOUND;
            serverResponse.statusMessage = HttpStatusMessages[HttpStatusCode.NOT_FOUND];
            serverResponse.end();
            return;
        }
        parsedUrl.path = relativePath;
        incomingMessage.originalUrl = originalUrl;
        incomingMessage.baseUrl = this.basePath.toString();
        incomingMessage.parsedUrl = parsedUrl;
        Object.defineProperty(incomingMessage, 'searchParams', {
            get() {
                return incomingMessage.parsedUrl.searchParams;
            }
        });
        const req = HttpServerRequest.from(incomingMessage);
        const res = HttpServerResponse.from(serverResponse);
        this.handleHttp(req, res)
            .catch((e) => this._logger.fatal(e));
    }
    static async create(api, options) {
        const adapter = new NodeHttpAdapterHost();
        await adapter.init(api, options);
        return adapter;
    }
}
