// valgen/esm/rules/is-string.js
import { validator } from '../core/index.js';
/**
 * Validates if value is "string".
 * Converts input value to string if coerce option is set to 'true'.
 * @validator isString
 */
export function isString(options) {
    return validator('isString', function (input, context, _this) {
        if (input != null && typeof input !== 'string' && context.coerce) {
            if (typeof input === 'object') {
                if (typeof input.toJSON === 'function')
                    input = input.toJSON();
                else
                    input = JSON.stringify(input);
            }
            else
                input = String(input);
        }
        if (typeof input === 'string')
            return input;
        context.fail(_this, `{{label}} must be a string`, input);
    }, options);
}
export function stringReplace(searchValue, replacer) {
    return validator('stringReplace', function (input) {
        if (input == null)
            return input;
        return String(input).replace(searchValue, replacer);
    });
}
export function stringSplit(splitter, limit) {
    return validator('stringSplit', function (input) {
        if (input == null)
            return input;
        return String(input).split(splitter, limit);
    });
}
/**
 * Removes whitespace from both ends of a string
 * @validator trim
 */
export function trim() {
    return validator('trim', function (input) {
        if (input == null)
            return input;
        return String(input).trim();
    });
}
// *************************************************************
/**
 * Removes whitespace from the end of a string
 * @validator trimEnd
 */
export const trimEnd = () => trimEndRule;
const trimEndRule = validator('trimEnd', function (input) {
    if (input == null)
        return input;
    return String(input).trimEnd();
});
// *************************************************************
/**
 * Removes whitespace from the beginning of a string
 * @validator trimStart
 */
export const trimStart = () => trimStartRule;
const trimStartRule = validator('trimStart', function (input) {
    if (input == null)
        return input;
    return String(input).trimStart();
});
