// @browsery/stream/esm/stream.mjs
import { D as DefaultStream, b as browserExports } from './readable-stream-8kla9gs4.mjs';
import 'buffer';
import 'abort-controller';
import 'process';
import 'events';
import 'string_decoder';



var Duplex = browserExports.Duplex;
var PassThrough = browserExports.PassThrough;
var Readable = browserExports.Readable;
var Stream = browserExports.Stream;
var Transform = browserExports.Transform;
var Writable = browserExports.Writable;
var _isUint8Array = browserExports._isUint8Array;
var _uint8ArrayToBuffer = browserExports._uint8ArrayToBuffer;
var addAbortSignal = browserExports.addAbortSignal;
var compose = browserExports.compose;
var destroy = browserExports.destroy;
var finished = browserExports.finished;
var isDisturbed = browserExports.isDisturbed;
var isErrored = browserExports.isErrored;
var isReadable = browserExports.isReadable;
var pipeline = browserExports.pipeline;
var promises = browserExports.promises;
export { Duplex, PassThrough, Readable, Stream, Transform, Writable, _isUint8Array, _uint8ArrayToBuffer, addAbortSignal, compose, DefaultStream as default, destroy, finished, isDisturbed, isErrored, isReadable, pipeline, promises };
