// @opra/common/cjs/schema/type-guards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isContainer = exports.isStorage = exports.isSingleton = exports.isCollection = exports.isResource = exports.isEnumType = exports.isMappedType = exports.isMixinType = exports.isSimpleType = exports.isComplexType = exports.isDataType = void 0;
const complex_type_interface_js_1 = require("./data-type/complex-type.interface.js");
const enum_type_interface_js_1 = require("./data-type/enum-type.interface.js");
const mapped_type_interface_js_1 = require("./data-type/mapped-type.interface.js");
const mixin_type_interface_js_1 = require("./data-type/mixin-type.interface.js");
const simple_type_interface_js_1 = require("./data-type/simple-type.interface.js");
const collection_interface_js_1 = require("./resource/collection.interface.js");
const container_interface_js_1 = require("./resource/container.interface.js");
const singleton_interface_js_1 = require("./resource/singleton.interface.js");
const storage_interface_js_1 = require("./resource/storage.interface.js");
function isDataType(obj) {
    return obj && typeof obj === 'object' &&
        (obj.kind === complex_type_interface_js_1.ComplexType.Kind ||
            obj.kind === enum_type_interface_js_1.EnumType.Kind ||
            obj.kind === mapped_type_interface_js_1.MappedType.Kind ||
            obj.kind === simple_type_interface_js_1.SimpleType.Kind ||
            obj.kind === mixin_type_interface_js_1.MixinType.Kind);
}
exports.isDataType = isDataType;
function isComplexType(obj) {
    return obj && typeof obj === 'object' && obj.kind === complex_type_interface_js_1.ComplexType.Kind;
}
exports.isComplexType = isComplexType;
function isSimpleType(obj) {
    return obj && typeof obj === 'object' && obj.kind === simple_type_interface_js_1.SimpleType.Kind;
}
exports.isSimpleType = isSimpleType;
function isMixinType(obj) {
    return obj && typeof obj === 'object' && obj.kind === mixin_type_interface_js_1.MixinType.Kind;
}
exports.isMixinType = isMixinType;
function isMappedType(obj) {
    return obj && typeof obj === 'object' && obj.kind === mapped_type_interface_js_1.MappedType.Kind;
}
exports.isMappedType = isMappedType;
function isEnumType(obj) {
    return obj && typeof obj === 'object' && obj.kind === enum_type_interface_js_1.EnumType.Kind;
}
exports.isEnumType = isEnumType;
function isResource(obj) {
    return obj && typeof obj === 'object' &&
        (obj.kind === container_interface_js_1.Container.Kind ||
            obj.kind === collection_interface_js_1.Collection.Kind ||
            obj.kind === singleton_interface_js_1.Singleton.Kind ||
            obj.kind === storage_interface_js_1.Storage.Kind);
}
exports.isResource = isResource;
function isCollection(obj) {
    return obj && typeof obj === 'object' && obj.kind === collection_interface_js_1.Collection.Kind;
}
exports.isCollection = isCollection;
function isSingleton(obj) {
    return obj && typeof obj === 'object' && obj.kind === singleton_interface_js_1.Singleton.Kind;
}
exports.isSingleton = isSingleton;
function isStorage(obj) {
    return obj && typeof obj === 'object' && obj.kind === storage_interface_js_1.Storage.Kind;
}
exports.isStorage = isStorage;
function isContainer(obj) {
    return obj && typeof obj === 'object' && obj.kind === container_interface_js_1.Container.Kind;
}
exports.isContainer = isContainer;
