// @opra/common/cjs/document/factory/api-document-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiDocumentFactory = void 0;
const index_js_1 = require("../../helpers/index.js");
const index_js_2 = require("../../schema/index.js");
const api_document_js_1 = require("../api-document.js");
const constants_js_1 = require("../constants.js");
const complex_type_js_1 = require("../data-type/complex-type.js");
const enum_type_js_1 = require("../data-type/enum-type.js");
const collection_js_1 = require("../resource/collection.js");
const container_js_1 = require("../resource/container.js");
const singleton_js_1 = require("../resource/singleton.js");
const storage_js_1 = require("../resource/storage.js");
const type_document_factory_js_1 = require("./type-document-factory.js");
/**
 * @class ApiDocumentFactory
 */
class ApiDocumentFactory extends type_document_factory_js_1.TypeDocumentFactory {
    constructor() {
        super(...arguments);
        this.resourceQueue = new index_js_1.ResponsiveMap();
    }
    /**
     * Creates ApiDocument instance from given schema object
     * @param init
     */
    static async createDocument(init) {
        const factory = new ApiDocumentFactory();
        const document = factory.document = new api_document_js_1.ApiDocument();
        await factory.initDocument(init);
        return document;
    }
    /**
     * Downloads schema from the given URL and creates the document instance   * @param url
     */
    static async createDocumentFromUrl(url) {
        const factory = new ApiDocumentFactory();
        const document = factory.document = new api_document_js_1.ApiDocument();
        await factory.initDocumentFromUrl(url);
        return document;
    }
    async initDocument(init) {
        await super.initDocument(init);
        const processContainer = async (container, containerInit) => {
            if (!containerInit.resources)
                return;
            if (Array.isArray(containerInit.resources)) {
                for (const thunk of containerInit.resources) {
                    const initArguments = await this.importResourceInstance(thunk);
                    container.resources.set(initArguments.name, await this.createResource(container, initArguments));
                }
            }
            else
                for (const [name, schema] of Object.entries(containerInit.resources)) {
                    const initArguments = await this.importResourceSchema(name, schema);
                    container.resources.set(initArguments.name, await this.createResource(container, initArguments));
                }
            container.resources.sort();
        };
        if (init.root) {
            this.curPath.push('/root');
            await processContainer(this.document.root, init.root);
            this.curPath.pop();
            this.document.invalidate();
        }
        return this.document;
    }
    async importResourceSchema(name, schema) {
        const convertEndpoints = async (source) => {
            if (!source)
                return;
            const output = {};
            for (const [endpointName, endpointSchema] of Object.entries(source)) {
                /* istanbul ignore next */
                if (!endpointSchema)
                    continue;
                const outputEndpoint = output[endpointName] = { ...endpointSchema };
                let parameters;
                // Resolve lazy type
                if (endpointSchema.returnType) {
                    outputEndpoint.returnType = await this.importDataType(endpointSchema.returnType);
                }
                if (endpointSchema.parameters) {
                    parameters = outputEndpoint.parameters = {};
                    for (const [kP, oP] of Object.entries(endpointSchema.parameters)) {
                        if (oP.enum) {
                            oP.type = (0, enum_type_js_1.EnumType)(oP.enum, { name: kP + 'Enum' });
                        }
                        parameters[kP] = {
                            ...oP,
                            type: await this.importDataType(oP.type || 'any')
                        };
                    }
                }
            }
            return output;
        };
        if (schema.kind === 'Collection') {
            return {
                ...schema,
                kind: schema.kind,
                name,
                type: await this.importDataType(schema.type),
                actions: await convertEndpoints(schema.actions),
                operations: await convertEndpoints(schema.operations),
            };
        }
        else if (schema.kind === 'Singleton') {
            return {
                ...schema,
                kind: schema.kind,
                name,
                type: await this.importDataType(schema.type),
                actions: await convertEndpoints(schema.actions),
                operations: await convertEndpoints(schema.operations),
            };
        }
        else if (schema.kind === 'Storage') {
            return {
                ...schema,
                name,
                actions: await convertEndpoints(schema.actions),
                operations: await convertEndpoints(schema.operations),
            };
        }
        else if (schema.kind === 'Container') {
            const resources = [];
            if (schema.resources) {
                for (const [k, o] of Object.entries(schema.resources)) {
                    const rinit = await this.importResourceSchema(k, o);
                    resources.push(rinit);
                }
            }
            return {
                ...schema,
                name,
                resources,
                actions: await convertEndpoints(schema.actions)
            };
        }
        throw new TypeError(`Can not import resource schema (${schema.kind})`);
    }
    async importResourceInstance(thunk) {
        thunk = await (0, index_js_1.resolveThunk)(thunk);
        let ctor;
        let metadata;
        let instance;
        if (typeof thunk === 'function') {
            ctor = thunk;
        }
        else {
            ctor = Object.getPrototypeOf(thunk).constructor;
            instance = thunk;
            if (!Reflect.hasMetadata(constants_js_1.RESOURCE_METADATA, ctor) &&
                index_js_2.OpraSchema.isResource(thunk) && typeof thunk.controller === 'object') {
                ctor = Object.getPrototypeOf(thunk.controller).constructor;
                metadata = thunk;
                instance = thunk.controller;
            }
        }
        metadata = metadata || Reflect.getMetadata(constants_js_1.RESOURCE_METADATA, ctor);
        if (!metadata && index_js_2.OpraSchema.isResource(metadata))
            throw new TypeError(`Class "${ctor.name}" doesn't have a valid Resource metadata`);
        const convertEndpoints = async (source) => {
            if (!source)
                return;
            const output = {};
            for (const [kA, oA] of Object.entries(source)) {
                const o = output[kA] = { ...oA };
                // Resolve lazy type
                if (oA.returnType) {
                    o.returnType = await this.importDataType(oA.returnType);
                }
                if (oA.parameters) {
                    const parameters = o.parameters = {};
                    for (const [kP, oP] of Object.entries(oA.parameters)) {
                        if (oP.enum) {
                            oP.type = (0, enum_type_js_1.EnumType)(oP.enum, { name: kP + 'Enum' });
                        }
                        parameters[kP] = {
                            ...oP,
                            type: await this.importDataType(oP.type || 'any')
                        };
                    }
                }
                if (oA.options?.inputOverwriteFields) {
                    const inputOverwriteFields = {};
                    for (const [kP, oP] of Object.entries(oA.options.inputOverwriteFields)) {
                        if (oP.enum) {
                            oP.type = (0, enum_type_js_1.EnumType)(oP.enum, { name: kP + 'Enum' });
                        }
                        inputOverwriteFields[kP] = { ...oP, };
                        if (oP.type)
                            inputOverwriteFields[kP].type = await this.importDataType(oP.type);
                    }
                    o.options.inputOverwriteFields = inputOverwriteFields;
                }
                if (oA.options?.outputOverwriteFields) {
                    const outputOverwriteFields = {};
                    for (const [kP, oP] of Object.entries(oA.options.outputOverwriteFields)) {
                        if (oP.enum) {
                            oP.type = (0, enum_type_js_1.EnumType)(oP.enum, { name: kP + 'Enum' });
                        }
                        outputOverwriteFields[kP] = { ...oP };
                        if (oP.type)
                            outputOverwriteFields[kP].type = await this.importDataType(oP.type);
                    }
                    o.options.outputOverwriteFields = outputOverwriteFields;
                }
            }
            return output;
        };
        // Clone metadata to prevent changing its contents
        const initArguments = (0, index_js_1.cloneObject)(metadata);
        initArguments.controller = instance;
        initArguments.ctor = ctor;
        if (initArguments.actions)
            initArguments.actions = await convertEndpoints(initArguments.actions);
        if (initArguments.kind === 'Collection' || initArguments.kind === 'Singleton') {
            const dataType = await this.importDataType((metadata).type);
            if (!dataType)
                throw new TypeError(`Unable to determine data type of "${initArguments.name}" resource`);
            if (!(dataType instanceof complex_type_js_1.ComplexType))
                throw new TypeError(`Data type of "${initArguments.name}" resource is not a ComplexType`);
            initArguments.type = dataType;
            if (initArguments.operations)
                initArguments.operations = await convertEndpoints(initArguments.operations);
        }
        else if (initArguments.kind === 'Container') {
            const oldResources = initArguments.resources;
            if (Array.isArray(oldResources)) {
                initArguments.resources = [];
                for (const t of oldResources) {
                    const rinit = await this.importResourceInstance(t);
                    initArguments.resources.push(rinit);
                }
            }
        }
        return initArguments;
    }
    async createResource(container, initArguments) {
        if (initArguments.kind === 'Collection')
            return new collection_js_1.Collection(container, initArguments);
        if (initArguments.kind === 'Singleton')
            return new singleton_js_1.Singleton(container, initArguments);
        if (initArguments.kind === 'Storage')
            return new storage_js_1.Storage(container, initArguments);
        if (initArguments.kind === 'Container') {
            const newContainer = new container_js_1.Container(container, { ...initArguments, resources: undefined });
            if (initArguments.resources) {
                for (const r of initArguments.resources) {
                    const res = await this.createResource(newContainer, r);
                    newContainer.resources.set(res.name, res);
                }
            }
            return newContainer;
        }
        else
            throw new Error(`Unknown resource type ${initArguments.kind}`);
    }
}
exports.ApiDocumentFactory = ApiDocumentFactory;
