// @opra/common/cjs/document/api-document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiDocument = void 0;
const container_js_1 = require("./resource/container.js");
const type_document_js_1 = require("./type-document.js");
class ApiDocument extends type_document_js_1.TypeDocument {
    constructor() {
        super(...arguments);
        this.root = new container_js_1.Container(this, { name: '' });
    }
    getResource(path, silent) {
        return this.root.getResource(path, silent);
    }
    getContainer(path, silent) {
        return this.root.getContainer(path, silent);
    }
    getCollection(path, silent) {
        return this.root.getCollection(path, silent);
    }
    getSingleton(path, silent) {
        return this.root.getSingleton(path, silent);
    }
    getStorage(path, silent) {
        return this.root.getStorage(path, silent);
    }
    /**
     * Export as Opra schema definition object
     */
    exportSchema(options) {
        const schema = super.exportSchema(options);
        schema.root = this.root.exportSchema(options);
        delete schema.root.kind;
        return schema;
    }
}
exports.ApiDocument = ApiDocument;
