// @opra/common/esm/document/resource/action.js
import { isAny } from 'valgen';
import { DataType } from '../data-type/data-type.js';
import { Endpoint } from './endpoint.js';
/**
 *
 * @class Action
 */
export class Action extends Endpoint {
    constructor(resource, name, init) {
        super(resource, name, init);
        this.resource = resource;
        this.name = name;
        this.kind = 'action';
        this.encodeReturning = isAny;
        if (init.returnType)
            this.returnType = init.returnType instanceof DataType
                ? init.returnType : this.resource.document.getDataType(init.returnType);
        this.returnMime = init.returnMime;
    }
    exportSchema(options) {
        const schema = super.exportSchema(options);
        if (this.returnType)
            schema.returnType = this.returnType.isEmbedded
                ? this.returnType.exportSchema(options)
                : this.returnType.name;
        if (this.returnMime)
            schema.returnMime = this.returnMime;
        return schema;
    }
}
