// @opra/common/esm/document/resource/endpoint.js
import { omitUndefined, ResponsiveMap } from '../../helpers/index.js';
import { DataType } from '../data-type/data-type.js';
import { Parameter } from './parameter.js';
/**
 *
 * @class Endpoint
 */
export class Endpoint {
    constructor(resource, name, init) {
        this.resource = resource;
        this.name = name;
        this.options = {};
        Object.assign(this, init);
        this.parameters = new ResponsiveMap();
        if (init.parameters) {
            for (const [n, p] of Object.entries(init.parameters)) {
                this.defineParameter(n, p);
            }
        }
    }
    getFullPath(documentPath) {
        return this.resource.getFullPath(documentPath) +
            (this.kind === 'action' ? (documentPath ? '/actions/' : '/') + this.name : '');
    }
    defineParameter(name, init) {
        const type = init.type && init.type instanceof DataType
            ? init.type : this.resource.document.getDataType(init.type || 'any');
        const prm = new Parameter(name, {
            ...init,
            type
        });
        this.parameters.set(prm.name, prm);
        return prm;
    }
    exportSchema(options) {
        const schema = omitUndefined({
            description: this.description
        });
        if (Object.keys(this.options).length)
            schema.options = { ...this.options };
        if (this.parameters.size) {
            let i = 0;
            const parameters = {};
            for (const [name, param] of this.parameters.entries()) {
                if (!param.isBuiltin) {
                    parameters[name] = param.exportSchema(options);
                    i++;
                }
            }
            if (i)
                schema.parameters = parameters;
        }
        return schema;
    }
}
