// @opra/common/esm/document/resource/parameter.js
import { vg } from 'valgen';
import { omitUndefined } from '../../helpers/index.js';
/**
 *
 * @class Parameter
 */
export class Parameter {
    constructor(name, init) {
        this.name = name;
        this.type = init.type;
        this.description = init.description;
        this.isArray = init.isArray;
        this.default = init.default;
        this.required = init.required;
        this.deprecated = init.deprecated;
        this.examples = init.examples;
        this.isBuiltin = init.isBuiltin;
    }
    exportSchema(options) {
        return omitUndefined({
            type: this.type.name ? this.type.name : this.type.exportSchema(options),
            description: this.description,
            isArray: this.isArray,
            default: this.default,
            required: this.required,
            deprecated: this.deprecated,
            examples: this.examples
        });
    }
    getDecoder() {
        if (!this._decoder)
            this._decoder = this.generateCodec('decode');
        return this._decoder;
    }
    getEncoder() {
        if (!this._encoder)
            this._encoder = this.generateCodec('encode');
        return this._encoder;
    }
    generateCodec(codec, options) {
        let fn = this.type.generateCodec(codec, options);
        if (this.isArray)
            fn = vg.stringSplit(',');
        return !options?.partial && this.required ? vg.required(fn) : vg.optional(fn);
    }
}
