// @opra/common/esm/document/resource/singleton-class.js
import { OpraSchema } from '../../schema/index.js';
import { CrudResource } from './crud-resource.js';
export class SingletonClass extends CrudResource {
    constructor(parent, init) {
        super(parent, init);
        this.kind = OpraSchema.Singleton.Kind;
        this.type = init.type;
        // ------------------
        let endpoint = this.operations.get('create');
        if (endpoint) {
            endpoint.defineParameter('pick', { type: 'string', isArray: true, isBuiltin: true });
            endpoint.defineParameter('omit', { type: 'string', isArray: true, isBuiltin: true });
            endpoint.defineParameter('include', { type: 'string', isArray: true, isBuiltin: true });
            endpoint.decodeInput = this.type.generateCodec('decode', {
                partial: true,
                pick: endpoint.options.inputPickFields,
                omit: endpoint.options.inputOmitFields,
                operation: 'write',
                overwriteFields: endpoint.inputOverwriteFields
            });
            endpoint.returnType = this.type;
            endpoint.encodeReturning = endpoint.returnType.generateCodec('encode', {
                partial: true,
                pick: endpoint.options.outputPickFields,
                omit: endpoint.options.outputOmitFields,
                operation: 'read',
                overwriteFields: endpoint.outputOverwriteFields
            });
        }
        // ------------------
        endpoint = this.operations.get('get');
        if (endpoint) {
            endpoint.defineParameter('pick', { type: 'string', isArray: true, isBuiltin: true });
            endpoint.defineParameter('omit', { type: 'string', isArray: true, isBuiltin: true });
            endpoint.defineParameter('include', { type: 'string', isArray: true, isBuiltin: true });
            endpoint.returnType = this.type;
            endpoint.encodeReturning = endpoint.returnType.generateCodec('encode', {
                partial: true,
                pick: endpoint.options.outputPickFields,
                omit: endpoint.options.outputOmitFields,
                operation: 'read',
                overwriteFields: endpoint.outputOverwriteFields
            });
        }
        // ------------------
        endpoint = this.operations.get('update');
        if (endpoint) {
            endpoint.defineParameter('pick', { type: 'string', isArray: true, isBuiltin: true });
            endpoint.defineParameter('omit', { type: 'string', isArray: true, isBuiltin: true });
            endpoint.defineParameter('include', { type: 'string', isArray: true, isBuiltin: true });
            endpoint.decodeInput = this.type.generateCodec('decode', {
                pick: endpoint.options.inputPickFields,
                omit: endpoint.options.inputOmitFields,
                operation: 'write',
                overwriteFields: endpoint.inputOverwriteFields
            });
            endpoint.returnType = this.type;
            endpoint.encodeReturning = endpoint.returnType.generateCodec('encode', {
                partial: true,
                pick: endpoint.options.outputPickFields,
                omit: endpoint.options.outputOmitFields,
                operation: 'read',
                overwriteFields: endpoint.outputOverwriteFields
            });
        }
    }
    getOperation(name) {
        return super.getOperation(name);
    }
    exportSchema(options) {
        return {
            ...super.exportSchema(options),
            type: this.type.name || 'any'
        };
    }
    normalizeFieldNames(path) {
        return this.type.normalizeFieldNames(path);
    }
}
