// @opra/core/cjs/services/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
class Logger {
    constructor(options = {}) {
        this._instance = options.instance ||
            (!(process.env.NODE_ENV || '').includes('test') ? globalThis.console : {});
    }
    info(message, ...optionalParams) {
        (this._instance.info || this._instance.log)?.(message, ...optionalParams);
    }
    error(message, ...optionalParams) {
        if (this._instance.error) {
            this._instance.error(message, ...optionalParams);
            return;
        }
        this.info(message, ...optionalParams);
    }
    fatal(message, ...optionalParams) {
        if (this._instance.fatal) {
            this._instance.fatal(message, ...optionalParams);
            return;
        }
        this.error(message, ...optionalParams);
    }
    warn(message, ...optionalParams) {
        this._instance.warn?.(message, ...optionalParams);
    }
    debug(message, ...optionalParams) {
        this._instance.debug?.(message, ...optionalParams);
    }
    verbose(message, ...optionalParams) {
        this._instance.verbose?.(message, ...optionalParams);
    }
}
exports.Logger = Logger;
