// @opra/common/esm/document/data-type/data-type.js
import { omitUndefined } from '../../helpers/index.js';
import { colorFgMagenta, colorFgYellow, colorReset, nodeInspectCustom } from '../utils/inspect.util.js';
/**
 * @class DataType
 * @abstract
 */
export class DataType {
    constructor(document, init) {
        this.document = document;
        this.name = init?.name;
        this.own = {};
        this.description = init?.description;
        if (!this.name)
            this.isEmbedded = true;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    exportSchema(options) {
        return omitUndefined({
            kind: this.kind,
            description: this.description
        });
    }
    extendsFrom(type) {
        const dataType = type instanceof DataType ? type : this.document.getDataType(type);
        let t = this;
        while (t) {
            if (t === dataType)
                return true;
            t = t.base;
        }
        return false;
    }
    toString() {
        return `[${Object.getPrototypeOf(this).constructor.name} ${this.name || '#Embedded'}]`;
    }
    [nodeInspectCustom]() {
        return `[${colorFgYellow + Object.getPrototypeOf(this).constructor.name + colorReset}` +
            ` ${colorFgMagenta + this.name + colorReset}]`;
    }
}
