// @opra/common/cjs/document/data-type/simple-type-class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleTypeClass = void 0;
const valgen_1 = require("valgen");
const index_js_1 = require("../../helpers/index.js");
const index_js_2 = require("../../schema/index.js");
const data_type_js_1 = require("./data-type.js");
/**
 * @class SimpleType
 */
class SimpleTypeClass extends data_type_js_1.DataType {
    constructor(document, init) {
        super(document, init);
        this.kind = index_js_2.OpraSchema.SimpleType.Kind;
        this.base = init.base;
        this.decode = init.decoder || init.base?.decode || valgen_1.isAny;
        this.encode = init.encoder || init.base?.encode || valgen_1.isAny;
    }
    exportSchema(options) {
        // noinspection UnnecessaryLocalVariableJS
        const out = super.exportSchema(options);
        Object.assign(out, (0, index_js_1.omitUndefined)({
            base: this.base ?
                (this.base.name ? this.base.name : this.base.exportSchema(options)) : undefined,
            decode: !options?.webSafe ? this.decode : undefined,
            encode: !options?.webSafe ? this.encode : undefined
        }));
        return out;
    }
    generateCodec(codec) {
        if (codec === 'encode')
            return this.encode;
        else
            return this.decode;
    }
}
exports.SimpleTypeClass = SimpleTypeClass;
